Refactor using a node block

glarizza opened the pull request:
This will allow everything to be performed on an executor

glarizza committed:
Remove unused ENV vars

glarizza committed:
Restructure and document the pipeline

This commit removes the explicit node block in favor of using the agent
block for determining the executor that will need to run the code.
An explicit git checkout stage has been added to ensure that the code is
checked out on the executor before anything happens. I'm unsure if the
checkout in the pipeline ONLY runs on the master or if it will be
triggered on the build executor too (if it runs on the executor then we
can strip this stage later - will need more testing). Finally
documentation has been added to the pipeline to make it easier to follow
and understand. We're dealing with a case of wanting to make sure a team
that's new to CI can follow what is going on, so documentation is
important.

glarizza commented:
Output of `terraform plan` within the repository's `project/intermediate/two` directory initiated from Jenkins job `Terraform Validation Pipeline/PR-10` build `18`:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_compute_address.external will be created
  + resource "google_compute_address" "external" {
      + address            = (known after apply)
      + address_type       = "EXTERNAL"
      + creation_timestamp = (known after apply)
      + id                 = (known after apply)
      + name               = "jenkins-testing"
      + network_tier       = (known after apply)
      + project            = "test-project"
      + region             = "us-west1"
      + self_link          = (known after apply)
      + subnetwork         = (known after apply)
      + users              = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

glarizza commented:
Output of `terraform plan` within the repository's `project/intermediate/two` directory initiated from Jenkins job `Terraform Validation Pipeline/PR-10` build `19`:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_compute_address.external will be created
  + resource "google_compute_address" "external" {
      + address            = (known after apply)
      + address_type       = "EXTERNAL"
      + creation_timestamp = (known after apply)
      + id                 = (known after apply)
      + name               = "jenkins-testing"
      + network_tier       = (known after apply)
      + project            = "test-project"
      + region             = "us-west1"
      + self_link          = (known after apply)
      + subnetwork         = (known after apply)
      + users              = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

glarizza commented:
Output of `terraform plan` within the repository's `project/intermediate/two` directory initiated from Jenkins job `Terraform Validation Pipeline/PR-10` build `22`:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_compute_address.external will be created
  + resource "google_compute_address" "external" {
      + address            = (known after apply)
      + address_type       = "EXTERNAL"
      + creation_timestamp = (known after apply)
      + id                 = (known after apply)
      + name               = "jenkins-testing"
      + network_tier       = (known after apply)
      + project            = "test-project"
      + region             = "us-west1"
      + self_link          = (known after apply)
      + subnetwork         = (known after apply)
      + users              = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

glarizza commented:
Output of `terraform plan` within the repository's `project/intermediate/two` directory initiated from Jenkins job `Terraform Validation Pipeline/PR-10` build `23`:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_compute_address.external will be created
  + resource "google_compute_address" "external" {
      + address            = (known after apply)
      + address_type       = "EXTERNAL"
      + creation_timestamp = (known after apply)
      + id                 = (known after apply)
      + name               = "jenkins-testing"
      + network_tier       = (known after apply)
      + project            = "test-project"
      + region             = "us-west1"
      + self_link          = (known after apply)
      + subnetwork         = (known after apply)
      + users              = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

glarizza commented:
Output of `terraform plan` within the repository's `project/intermediate/two` directory initiated from Jenkins job `Terraform Validation Pipeline/PR-10` build `24`:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_compute_address.external will be created
  + resource "google_compute_address" "external" {
      + address            = (known after apply)
      + address_type       = "EXTERNAL"
      + creation_timestamp = (known after apply)
      + id                 = (known after apply)
      + name               = "jenkins-testing"
      + network_tier       = (known after apply)
      + project            = "test-project"
      + region             = "us-west1"
      + self_link          = (known after apply)
      + subnetwork         = (known after apply)
      + users              = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

glarizza commented:
Output of `terraform plan` within the repository's `project/intermediate/two` directory initiated from Jenkins job `Terraform Validation Pipeline/PR-10` build `26`:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_compute_address.external will be created
  + resource "google_compute_address" "external" {
      + address            = (known after apply)
      + address_type       = "EXTERNAL"
      + creation_timestamp = (known after apply)
      + id                 = (known after apply)
      + name               = "jenkins-testing"
      + network_tier       = (known after apply)
      + project            = "test-project"
      + region             = "us-west1"
      + self_link          = (known after apply)
      + subnetwork         = (known after apply)
      + users              = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

glarizza commented:
Output of `terraform plan` within the repository's `project/intermediate/two` directory initiated from Jenkins job `Terraform Validation Pipeline/PR-10` build `27`:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_compute_address.external will be created
  + resource "google_compute_address" "external" {
      + address            = (known after apply)
      + address_type       = "EXTERNAL"
      + creation_timestamp = (known after apply)
      + id                 = (known after apply)
      + name               = "jenkins-testing"
      + network_tier       = (known after apply)
      + project            = "test-project"
      + region             = "us-west1"
      + self_link          = (known after apply)
      + subnetwork         = (known after apply)
      + users              = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

ghoneycutt commented on Jenkinsfile:
```diff
@@ -13,6 +17,14 @@ def getChangedFiles() {
+     Arguments:
+       mappingsFilePath: The path to the YAML mapping file
+
+     Returns: A map of the YAML file with each directory being a key in the map
```
This is now a list instead of a map. Other s/map/list or other rewording around that to be done

glarizza replied:
Not in my repo it isn't :) @ghoneycutt can you get the list of directory paths committed into this repo so I can test it out and update my code to reflect it?

glarizza replied:
This should be resolved now - updated the mapping file and code paths

ghoneycutt commented on Jenkinsfile:
```diff
@@ -22,24 +34,49 @@ def readTerraformMappingsFile(mappingsFilePath) {
+       terraformDirectoriesList: A list of relative paths in the repo that
+         correspond to individual terraform implementation directories.
+
+      Returns: The relative path to the terraform implementation directory
```
Path to the terraform implementation directory relative to the root of the repository.

glarizza replied:
Updated the wording

ghoneycutt commented on Jenkinsfile:
```diff
@@ -22,24 +34,49 @@ def readTerraformMappingsFile(mappingsFilePath) {
+       terraformDirectoriesList: A list of relative paths in the repo that
+         correspond to individual terraform implementation directories.
+
+      Returns: The relative path to the terraform implementation directory
```
probably should have some terminology defined at the top such as terraform implementation directory is a directory in which terraform is run.

glarizza replied:
I think that should be defined in documentation for the Pipeline, and not necessarily within the pipeline itself. I'm being mindful of having a ton of comments in the file to the point where people start ignoring them

ghoneycutt commented on Jenkinsfile:
```diff
@@ -22,24 +34,49 @@ def readTerraformMappingsFile(mappingsFilePath) {
+  // of an implementation directory have been changed)
+  terraformRootDirectoryList.unique()
+
+  // If file changes weren't in a terraform implementation directory return null
```
have the language match the code, so 

return null if blah blah, else continue.

also, what happens when we return null? is that just saying that there is nothing to do on the PR?

glarizza replied:
Updated the wording

ghoneycutt commented on Jenkinsfile:
```diff
@@ -53,72 +90,106 @@ def getTerraformRootDirectory(changedFilesList, terraformDirectoriesList) {
+      steps {
+        script {
+          dir(GIT_ROOT) {
+            def terraformDirectoryMap = readTerraformMappingsFile("${env.WORKSPACE}/terraform-directory-mappings.yaml")
```
`.terraform_directories.yaml`

glarizza replied:
Not in this branch it isn't - can you get that file committed so I can update the code?

glarizza replied:
This should be resolved now - updated the mapping file and code paths

ghoneycutt commented on Jenkinsfile:
```diff
@@ -53,72 +90,106 @@ def getTerraformRootDirectory(changedFilesList, terraformDirectoriesList) {
+        script {
+          dir(GIT_ROOT) {
+            def terraformDirectoryMap = readTerraformMappingsFile("${env.WORKSPACE}/terraform-directory-mappings.yaml")
+            def terraformDirectoriesList = terraformDirectoryMap.keySet()
```
what is `keySet()`?

should this method have a `get` in front of it to match the other methods?

glarizza replied:
`keySet()` == `.keys`, it's getting the keys to the map/hash that my `terraform-directory-mappings.yaml` file returns. It will be moot once the new mapping file is committed, but I need that to happen first before I change the code in my branch

glarizza replied:
This should be resolved now - updated the mapping file and code paths

ghoneycutt commented on Jenkinsfile:
```diff
@@ -53,72 +90,106 @@ def getTerraformRootDirectory(changedFilesList, terraformDirectoriesList) {
+    }
+
+    stage("Validate Terraform Directory") {
+      when {
```
if the stage it skipped, it should output that to the log.

glarizza replied:
It does, see the log for [Build #16](http://35.197.39.34:8080/job/Terraform%20Validation%20Pipeline/view/change-requests/job/PR-10/16/console)

```
[Pipeline] echo
No Terraform directories were modified in this PR. Exiting...
[Pipeline] }
[Pipeline] // dir
[Pipeline] }
[Pipeline] // script
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (Validate Terraform Directory)
Stage "Validate Terraform Directory" skipped due to when conditional
```

ghoneycutt commented on Jenkinsfile:
```diff
@@ -53,72 +90,106 @@ def getTerraformRootDirectory(changedFilesList, terraformDirectoriesList) {
+
+    stage("Validate Terraform Directory") {
+      when {
+        expression {
```
should explain the intent behind this is that if a change does not touch terraform code, then we skip and mark the build as success.

glarizza replied:
Updated the wording

ghoneycutt commented on Jenkinsfile:
```diff
@@ -53,72 +90,106 @@ def getTerraformRootDirectory(changedFilesList, terraformDirectoriesList) {
+            dir = terraformPath
+          }
+
+          if (env.CHANGE_ID) {
```
note what this `CHANGE_ID` is used for

glarizza replied:
Updated

ghoneycutt commented on Jenkinsfile:
```diff
@@ -53,72 +90,106 @@ def getTerraformRootDirectory(changedFilesList, terraformDirectoriesList) {
       }
     }
+
+    stage('Cleanup') {
```
would ditch the cleanup stage

glarizza replied:
Makes sense - removed it

glarizza committed:
Update terraform mappings workflow

The structure and location of the terraform mappings file has changed,
and this commit tracks those changes both in code as well as within the
repository. We're using a list instead of a map for the file's contents
because all we need right now are the paths to the implementation
directories.

glarizza commented:
Output of `terraform plan` within the repository's `project/intermediate/two` directory initiated from Jenkins job `Terraform Validation Pipeline/PR-10` build `29`:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_compute_address.external will be created
  + resource "google_compute_address" "external" {
      + address            = (known after apply)
      + address_type       = "EXTERNAL"
      + creation_timestamp = (known after apply)
      + id                 = (known after apply)
      + name               = "jenkins-testing"
      + network_tier       = (known after apply)
      + project            = "test-project"
      + region             = "us-west1"
      + self_link          = (known after apply)
      + subnetwork         = (known after apply)
      + users              = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

glarizza commented:
Output of `terraform plan` within the repository's `project/intermediate/two` directory initiated from Jenkins job `Terraform Validation Pipeline/PR-10` build `30`:
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # google_compute_address.external will be created
  + resource "google_compute_address" "external" {
      + address            = (known after apply)
      + address_type       = "EXTERNAL"
      + creation_timestamp = (known after apply)
      + id                 = (known after apply)
      + name               = "jenkins-testing"
      + network_tier       = (known after apply)
      + project            = "test-project"
      + region             = "us-west1"
      + self_link          = (known after apply)
      + subnetwork         = (known after apply)
      + users              = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```
